File: python/ivit/utils/visualizer.py
"""
Visualization utilities.
"""

from typing import List, Dict, Tuple, Optional
import numpy as np

from ..core.types import Detection, ClassificationResult, Pose


class Visualizer:
    """
    Visualization utility class.

    Provides static methods for drawing inference results on images.
    """

    # Default colors (BGR format)
    COLORS = [
        (255, 0, 0), (0, 255, 0), (0, 0, 255),
        (255, 255, 0), (255, 0, 255), (0, 255, 255),
        (128, 0, 0), (0, 128, 0), (0, 0, 128),
        (128, 128, 0), (128, 0, 128), (0, 128, 128),
        (255, 128, 0), (255, 0, 128), (128, 255, 0),
        (0, 255, 128), (128, 0, 255), (0, 128, 255),
    ]

    @staticmethod
    def draw_detections(
        image: np.ndarray,
        detections: List[Detection],
        labels: Optional[List[str]] = None,
        show_labels: bool = True,
        show_confidence: bool = True,
        thickness: int = 2,
        font_scale: float = 0.5,
    ) -> np.ndarray:
        """
        Draw detection boxes on image.

        Args:
            image: Input image (will be copied)
            detections: List of detections
            labels: Class labels
            show_labels: Show class names
            show_confidence: Show confidence scores
            thickness: Box line thickness
            font_scale: Font scale

        Returns:
            Image with drawn detections
        """
        import cv2

        vis = image.copy()

        for det in detections:
            color = Visualizer.get_color(det.class_id)
            x1, y1 = int(det.bbox.x1), int(det.bbox.y1)
            x2, y2 = int(det.bbox.x2), int(det.bbox.y2)

            # Draw box
            cv2.rectangle(vis, (x1, y1), (x2, y2), color, thickness)

            # Draw label
            if show_labels or show_confidence:
                label_parts = []
                if show_labels:
                    if labels and det.class_id < len(labels):
                        label_parts.append(labels[det.class_id])
                    else:
                        label_parts.append(det.label)
                if show_confidence:
                    label_parts.append(f"{det.confidence:.2f}")

                label = " ".join(label_parts)

                (tw, th), _ = cv2.getTextSize(
                    label, cv2.FONT_HERSHEY_SIMPLEX, font_scale, 1
                )
                cv2.rectangle(vis, (x1, y1 - th - 4), (x1 + tw, y1), color, -1)
                cv2.putText(
                    vis, label, (x1, y1 - 2),
                    cv2.FONT_HERSHEY_SIMPLEX, font_scale, (255, 255, 255), 1
                )

            # Draw mask if available
            if det.mask is not None:
                mask_color = np.array(color, dtype=np.float32)
                mask_overlay = (det.mask[:, :, np.newaxis] * mask_color).astype(np.uint8)
                vis = cv2.addWeighted(vis, 1, mask_overlay, 0.5, 0)

        return vis

    @staticmethod
    def draw_segmentation(
        image: np.ndarray,
        mask: np.ndarray,
        colormap: Optional[Dict[int, Tuple[int, int, int]]] = None,
        alpha: float = 0.5,
    ) -> np.ndarray:
        """
        Draw segmentation mask overlay.

        Args:
            image: Input image
            mask: Segmentation mask (H, W)
            colormap: Color mapping {class_id: (B, G, R)}
            alpha: Overlay transparency

        Returns:
            Image with mask overlay
        """
        import cv2

        vis = image.copy()
        h, w = mask.shape[:2]

        # Generate colormap if not provided
        if colormap is None:
            colormap = Visualizer.generate_colormap(int(mask.max()) + 1)

        # Create colored mask
        colored = np.zeros((h, w, 3), dtype=np.uint8)
        for class_id, color in colormap.items():
            colored[mask == class_id] = color

        # Blend
        vis = cv2.addWeighted(vis, 1 - alpha, colored, alpha, 0)

        return vis

    @staticmethod
    def draw_poses(
        image: np.ndarray,
        poses: List[Pose],
        skeleton: Optional[List[Tuple[int, int]]] = None,
        keypoint_radius: int = 3,
        thickness: int = 2,
    ) -> np.ndarray:
        """
        Draw pose keypoints and skeleton.

        Args:
            image: Input image
            poses: List of poses
            skeleton: Skeleton connections [(i, j), ...]
            keypoint_radius: Keypoint circle radius
            thickness: Line thickness

        Returns:
            Image with drawn poses
        """
        import cv2

        vis = image.copy()

        # Default COCO skeleton
        if skeleton is None:
            skeleton = [
                (0, 1), (0, 2), (1, 3), (2, 4),  # Head
                (5, 6), (5, 7), (7, 9), (6, 8), (8, 10),  # Arms
                (5, 11), (6, 12), (11, 12),  # Torso
                (11, 13), (13, 15), (12, 14), (14, 16),  # Legs
            ]

        for pose in poses:
            # Draw keypoints
            for i, kp in enumerate(pose.keypoints):
                if kp.confidence > 0.5:
                    color = Visualizer.get_color(i)
                    cv2.circle(
                        vis,
                        (int(kp.x), int(kp.y)),
                        keypoint_radius,
                        color,
                        -1
                    )

            # Draw skeleton
            for i, j in skeleton:
                if i < len(pose.keypoints) and j < len(pose.keypoints):
                    kp1, kp2 = pose.keypoints[i], pose.keypoints[j]
                    if kp1.confidence > 0.5 and kp2.confidence > 0.5:
                        cv2.line(
                            vis,
                            (int(kp1.x), int(kp1.y)),
                            (int(kp2.x), int(kp2.y)),
                            Visualizer.get_color(i),
                            thickness
                        )

        return vis

    @staticmethod
    def draw_classification(
        image: np.ndarray,
        results: List[ClassificationResult],
        top_k: int = 5,
        font_scale: float = 0.6,
    ) -> np.ndarray:
        """
        Draw classification results as text.

        Args:
            image: Input image
            results: Classification results
            top_k: Number of results to show
            font_scale: Font scale

        Returns:
            Image with classification text
        """
        import cv2

        vis = image.copy()

        for i, result in enumerate(results[:top_k]):
            text = f"{result.label}: {result.score:.2%}"
            y = 30 + i * 25

            cv2.putText(
                vis, text, (10, y),
                cv2.FONT_HERSHEY_SIMPLEX, font_scale,
                (0, 0, 0), 3  # Black outline
            )
            cv2.putText(
                vis, text, (10, y),
                cv2.FONT_HERSHEY_SIMPLEX, font_scale,
                (255, 255, 255), 1  # White text
            )

        return vis

    @staticmethod
    def create_comparison(
        images: List[np.ndarray],
        titles: Optional[List[str]] = None,
        cols: int = 2,
    ) -> np.ndarray:
        """
        Create side-by-side comparison image.

        Args:
            images: List of images
            titles: Titles for each image
            cols: Number of columns

        Returns:
            Combined image
        """
        import cv2

        n = len(images)
        rows = (n + cols - 1) // cols

        # Get max dimensions
        max_h = max(img.shape[0] for img in images)
        max_w = max(img.shape[1] for img in images)

        # Create canvas
        canvas = np.zeros((rows * (max_h + 30), cols * max_w, 3), dtype=np.uint8)

        for i, img in enumerate(images):
            row, col = i // cols, i % cols
            y, x = row * (max_h + 30) + 30, col * max_w

            # Resize if needed
            if img.shape[:2] != (max_h, max_w):
                img = cv2.resize(img, (max_w, max_h))

            canvas[y:y+max_h, x:x+max_w] = img

            # Add title
            if titles and i < len(titles):
                cv2.putText(
                    canvas, titles[i], (x + 10, y - 10),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 1
                )

        return canvas

    @staticmethod
    def get_color(class_id: int) -> Tuple[int, int, int]:
        """Get color for class ID."""
        return Visualizer.COLORS[class_id % len(Visualizer.COLORS)]

    @staticmethod
    def generate_colormap(num_classes: int) -> Dict[int, Tuple[int, int, int]]:
        """Generate colormap for given number of classes."""
        import colorsys

        colormap = {}
        for i in range(num_classes):
            hue = i / num_classes
            rgb = colorsys.hsv_to_rgb(hue, 0.8, 0.8)
            colormap[i] = tuple(int(c * 255) for c in rgb)

        # Background is black
        colormap[0] = (0, 0, 0)

        return colormap

File: src/utils/visualizer.cpp
/**
 * @file visualizer.cpp
 * @brief Visualization utilities implementation
 */

#include "ivit/utils/visualizer.hpp"
#include <opencv2/imgproc.hpp>
#include <sstream>
#include <iomanip>
#include <cmath>

namespace ivit {
namespace utils {

// ============================================================================
// Color palette (20 distinct colors for visualization)
// ============================================================================

const std::vector<cv::Scalar> Visualizer::COLORS = {
    {56, 56, 255},     // Red
    {50, 205, 50},     // Lime Green
    {255, 144, 30},    // Dodger Blue
    {0, 215, 255},     // Gold
    {255, 0, 255},     // Magenta
    {255, 255, 0},     // Cyan
    {0, 128, 128},     // Olive
    {128, 0, 128},     // Purple
    {0, 165, 255},     // Orange
    {203, 192, 255},   // Pink
    {128, 128, 0},     // Teal
    {60, 20, 220},     // Crimson
    {230, 216, 173},   // Light Steel Blue
    {144, 238, 144},   // Light Green
    {147, 112, 219},   // Medium Purple
    {255, 182, 193},   // Light Pink
    {0, 100, 0},       // Dark Green
    {139, 0, 0},       // Dark Blue
    {0, 139, 139},     // Dark Cyan
    {169, 169, 169},   // Dark Gray
};

// ============================================================================
// Visualizer implementation
// ============================================================================

cv::Scalar Visualizer::get_color(int class_id) {
    return COLORS[class_id % COLORS.size()];
}

std::map<int, cv::Vec3b> Visualizer::generate_colormap(int num_classes) {
    std::map<int, cv::Vec3b> colormap;

    // Background is always black
    colormap[0] = cv::Vec3b(0, 0, 0);

    for (int i = 1; i < num_classes; i++) {
        // Generate distinct colors using HSV color space
        float hue = static_cast<float>(i - 1) / static_cast<float>(num_classes - 1) * 180.0f;
        float saturation = 200.0f + (i % 3) * 20.0f;  // Vary saturation slightly
        float value = 180.0f + (i % 5) * 15.0f;       // Vary brightness slightly

        cv::Mat hsv(1, 1, CV_8UC3, cv::Scalar(hue, saturation, value));
        cv::Mat bgr;
        cv::cvtColor(hsv, bgr, cv::COLOR_HSV2BGR);
        colormap[i] = bgr.at<cv::Vec3b>(0, 0);
    }

    return colormap;
}

cv::Mat Visualizer::draw_detections(
    const cv::Mat& image,
    const std::vector<Detection>& detections,
    const std::vector<std::string>& labels,
    const VisConfig& config
) {
    cv::Mat output = image.clone();

    for (const auto& det : detections) {
        cv::Scalar color = get_color(det.class_id);

        // Get label text
        std::string label_text;
        if (config.show_labels) {
            if (!det.label.empty()) {
                label_text = det.label;
            } else if (det.class_id >= 0 &&
                       static_cast<size_t>(det.class_id) < labels.size()) {
                label_text = labels[det.class_id];
            } else {
                label_text = "class_" + std::to_string(det.class_id);
            }
        }

        if (config.show_confidence) {
            std::ostringstream ss;
            ss << std::fixed << std::setprecision(2) << det.confidence;
            if (!label_text.empty()) {
                label_text += " " + ss.str();
            } else {
                label_text = ss.str();
            }
        }

        // Draw bounding box
        if (config.show_boxes) {
            cv::Point pt1(static_cast<int>(det.bbox.x1), static_cast<int>(det.bbox.y1));
            cv::Point pt2(static_cast<int>(det.bbox.x2), static_cast<int>(det.bbox.y2));
            cv::rectangle(output, pt1, pt2, color, config.thickness);
        }

        // Draw label background and text
        if (!label_text.empty()) {
            int baseline = 0;
            cv::Size text_size = cv::getTextSize(
                label_text,
                cv::FONT_HERSHEY_SIMPLEX,
                config.font_scale,
                1,
                &baseline
            );

            int label_x = static_cast<int>(det.bbox.x1);
            int label_y = static_cast<int>(det.bbox.y1) - text_size.height - 4;

            // Ensure label stays within image bounds
            label_y = std::max(text_size.height + 4, label_y);
            label_x = std::max(0, label_x);

            // Draw label background
            cv::rectangle(
                output,
                cv::Point(label_x, label_y - text_size.height - 4),
                cv::Point(label_x + text_size.width + 4, label_y + 4),
                color,
                -1  // Filled
            );

            // Draw label text
            cv::putText(
                output,
                label_text,
                cv::Point(label_x + 2, label_y),
                cv::FONT_HERSHEY_SIMPLEX,
                config.font_scale,
                cv::Scalar(255, 255, 255),
                1,
                cv::LINE_AA
            );
        }

        // Draw instance mask if available
        if (det.mask.has_value() && config.show_masks) {
            cv::Mat mask = det.mask.value();
            cv::Mat mask_resized;

            // Resize mask to match bounding box if needed
            int box_w = static_cast<int>(det.bbox.width());
            int box_h = static_cast<int>(det.bbox.height());

            if (mask.size() != cv::Size(box_w, box_h)) {
                cv::resize(mask, mask_resized, cv::Size(box_w, box_h), 0, 0, cv::INTER_LINEAR);
            } else {
                mask_resized = mask;
            }

            // Create colored mask
            cv::Mat colored_mask(output.size(), CV_8UC3, cv::Scalar(0, 0, 0));
            cv::Rect roi(
                static_cast<int>(det.bbox.x1),
                static_cast<int>(det.bbox.y1),
                box_w,
                box_h
            );

            // Clip ROI to image bounds
            roi &= cv::Rect(0, 0, output.cols, output.rows);
            if (roi.width > 0 && roi.height > 0) {
                cv::Mat roi_mask = mask_resized(cv::Rect(0, 0, roi.width, roi.height));
                cv::Mat roi_colored = colored_mask(roi);

                for (int y = 0; y < roi.height; y++) {
                    for (int x = 0; x < roi.width; x++) {
                        if (roi_mask.at<uchar>(y, x) > 127) {
                            roi_colored.at<cv::Vec3b>(y, x) = cv::Vec3b(
                                static_cast<uchar>(color[0]),
                                static_cast<uchar>(color[1]),
                                static_cast<uchar>(color[2])
                            );
                        }
                    }
                }

                cv::addWeighted(output, 1.0, colored_mask, config.mask_alpha, 0, output);
            }
        }
    }

    return output;
}

cv::Mat Visualizer::draw_segmentation(
    const cv::Mat& image,
    const cv::Mat& mask,
    const std::map<int, cv::Vec3b>& colormap,
    float alpha
) {
    if (mask.empty()) {
        return image.clone();
    }

    cv::Mat output = image.clone();
    cv::Mat colored_mask(image.size(), CV_8UC3, cv::Scalar(0, 0, 0));

    // Resize mask if needed
    cv::Mat resized_mask;
    if (mask.size() != image.size()) {
        cv::resize(mask, resized_mask, image.size(), 0, 0, cv::INTER_NEAREST);
    } else {
        resized_mask = mask;
    }

    // Generate colormap if not provided
    std::map<int, cv::Vec3b> cmap = colormap;
    if (cmap.empty()) {
        double min_val, max_val;
        cv::minMaxLoc(resized_mask, &min_val, &max_val);
        cmap = generate_colormap(static_cast<int>(max_val) + 1);
    }

    // Apply colormap
    for (int y = 0; y < resized_mask.rows; y++) {
        for (int x = 0; x < resized_mask.cols; x++) {
            int class_id;
            if (resized_mask.type() == CV_32S) {
                class_id = resized_mask.at<int>(y, x);
            } else if (resized_mask.type() == CV_8U) {
                class_id = resized_mask.at<uchar>(y, x);
            } else {
                class_id = static_cast<int>(resized_mask.at<float>(y, x));
            }

            if (cmap.count(class_id)) {
                colored_mask.at<cv::Vec3b>(y, x) = cmap.at(class_id);
            }
        }
    }

    // Blend with original image
    cv::addWeighted(output, 1.0 - alpha, colored_mask, alpha, 0, output);

    return output;
}

cv::Mat Visualizer::draw_poses(
    const cv::Mat& image,
    const std::vector<Pose>& poses,
    const std::vector<std::pair<int, int>>& skeleton,
    const VisConfig& config
) {
    cv::Mat output = image.clone();

    // Default COCO skeleton if not provided
    std::vector<std::pair<int, int>> skel = skeleton;
    if (skel.empty()) {
        skel = {
            {0, 1}, {0, 2}, {1, 3}, {2, 4},              // Head
            {5, 6},                                       // Shoulders
            {5, 7}, {7, 9}, {6, 8}, {8, 10},              // Arms
            {5, 11}, {6, 12}, {11, 12},                   // Torso
            {11, 13}, {13, 15}, {12, 14}, {14, 16}        // Legs
        };
    }

    // Limb colors (for visual distinction)
    std::vector<cv::Scalar> limb_colors = {
        {0, 215, 255},   // Head - gold
        {0, 255, 0},     // Shoulders - green
        {255, 128, 0},   // Left arm - blue
        {0, 128, 255},   // Right arm - orange
        {255, 0, 255},   // Torso - magenta
        {255, 0, 0},     // Left leg - blue
        {0, 0, 255}      // Right leg - red
    };

    for (const auto& pose : poses) {
        // Draw bounding box if available
        if (pose.bbox.has_value() && config.show_boxes) {
            const auto& box = pose.bbox.value();
            cv::rectangle(
                output,
                cv::Point(static_cast<int>(box.x1), static_cast<int>(box.y1)),
                cv::Point(static_cast<int>(box.x2), static_cast<int>(box.y2)),
                cv::Scalar(0, 255, 0),
                config.thickness
            );
        }

        // Draw skeleton connections
        int limb_idx = 0;
        for (const auto& [i, j] : skel) {
            if (i >= static_cast<int>(pose.keypoints.size()) ||
                j >= static_cast<int>(pose.keypoints.size())) {
                continue;
            }

            const auto& kp1 = pose.keypoints[i];
            const auto& kp2 = pose.keypoints[j];

            float conf_threshold = 0.3f;
            if (kp1.confidence > conf_threshold && kp2.confidence > conf_threshold) {
                cv::Scalar color = limb_colors[limb_idx % limb_colors.size()];
                cv::line(
                    output,
                    cv::Point(static_cast<int>(kp1.x), static_cast<int>(kp1.y)),
                    cv::Point(static_cast<int>(kp2.x), static_cast<int>(kp2.y)),
                    color,
                    config.thickness,
                    cv::LINE_AA
                );
            }
            limb_idx++;
        }

        // Draw keypoints
        for (size_t i = 0; i < pose.keypoints.size(); i++) {
            const auto& kp = pose.keypoints[i];
            if (kp.confidence > 0.3f) {
                cv::Scalar color = get_color(static_cast<int>(i));

                // Draw filled circle
                cv::circle(
                    output,
                    cv::Point(static_cast<int>(kp.x), static_cast<int>(kp.y)),
                    4,
                    color,
                    -1,
                    cv::LINE_AA
                );

                // Draw border
                cv::circle(
                    output,
                    cv::Point(static_cast<int>(kp.x), static_cast<int>(kp.y)),
                    4,
                    cv::Scalar(0, 0, 0),
                    1,
                    cv::LINE_AA
                );

                // Draw label if enabled
                if (config.show_labels && !kp.name.empty()) {
                    cv::putText(
                        output,
                        kp.name,
                        cv::Point(static_cast<int>(kp.x) + 5, static_cast<int>(kp.y) - 5),
                        cv::FONT_HERSHEY_SIMPLEX,
                        config.font_scale * 0.7f,
                        cv::Scalar(255, 255, 255),
                        1,
                        cv::LINE_AA
                    );
                }
            }
        }

        // Draw confidence score
        if (config.show_confidence && pose.bbox.has_value()) {
            std::ostringstream ss;
            ss << std::fixed << std::setprecision(2) << pose.confidence;
            cv::putText(
                output,
                ss.str(),
                cv::Point(
                    static_cast<int>(pose.bbox.value().x1),
                    static_cast<int>(pose.bbox.value().y1) - 5
                ),
                cv::FONT_HERSHEY_SIMPLEX,
                config.font_scale,
                cv::Scalar(0, 255, 0),
                1,
                cv::LINE_AA
            );
        }
    }

    return output;
}

cv::Mat Visualizer::draw_classification(
    const cv::Mat& image,
    const std::vector<ClassificationResult>& results,
    int top_k
) {
    cv::Mat output = image.clone();

    int n = std::min(top_k, static_cast<int>(results.size()));
    int y_offset = 30;
    int line_height = 25;

    // Draw semi-transparent background
    int bg_height = n * line_height + 20;
    int bg_width = 300;
    cv::Rect bg_rect(10, 10, bg_width, bg_height);
    bg_rect &= cv::Rect(0, 0, output.cols, output.rows);

    if (bg_rect.width > 0 && bg_rect.height > 0) {
        cv::Mat roi = output(bg_rect);
        cv::Mat bg(roi.size(), roi.type(), cv::Scalar(0, 0, 0));
        cv::addWeighted(roi, 0.5, bg, 0.5, 0, roi);
    }

    // Draw results
    for (int i = 0; i < n; i++) {
        const auto& result = results[i];

        // Format text
        std::ostringstream ss;
        ss << std::fixed << std::setprecision(1);
        ss << (i + 1) << ". ";
        if (!result.label.empty()) {
            ss << result.label;
        } else {
            ss << "class_" << result.class_id;
        }
        ss << " (" << (result.score * 100.0f) << "%)";

        // Color based on rank
        cv::Scalar color;
        if (i == 0) {
            color = cv::Scalar(0, 255, 0);   // Green for top-1
        } else if (i == 1) {
            color = cv::Scalar(0, 200, 255); // Orange for top-2
        } else {
            color = cv::Scalar(200, 200, 200); // Gray for others
        }

        cv::putText(
            output,
            ss.str(),
            cv::Point(20, y_offset + i * line_height),
            cv::FONT_HERSHEY_SIMPLEX,
            0.6,
            color,
            1,
            cv::LINE_AA
        );

        // Draw confidence bar
        int bar_x = 20;
        int bar_y = y_offset + i * line_height + 5;
        int bar_width = static_cast<int>(result.score * 200);
        int bar_height = 4;

        cv::rectangle(
            output,
            cv::Point(bar_x, bar_y),
            cv::Point(bar_x + bar_width, bar_y + bar_height),
            color,
            -1
        );
    }

    return output;
}

cv::Mat Visualizer::create_comparison(
    const std::vector<cv::Mat>& images,
    const std::vector<std::string>& titles,
    int cols
) {
    if (images.empty()) {
        return cv::Mat();
    }

    // Determine grid size
    int n = static_cast<int>(images.size());
    int rows = (n + cols - 1) / cols;

    // Find max size
    int max_w = 0, max_h = 0;
    for (const auto& img : images) {
        max_w = std::max(max_w, img.cols);
        max_h = std::max(max_h, img.rows);
    }

    // Add space for titles
    int title_height = 30;
    int cell_h = max_h + title_height;
    int cell_w = max_w;

    // Create output canvas
    cv::Mat output(rows * cell_h, cols * cell_w, CV_8UC3, cv::Scalar(40, 40, 40));

    for (int i = 0; i < n; i++) {
        int row = i / cols;
        int col = i % cols;

        int x = col * cell_w;
        int y = row * cell_h;

        // Draw title
        if (static_cast<size_t>(i) < titles.size() && !titles[i].empty()) {
            cv::putText(
                output,
                titles[i],
                cv::Point(x + 10, y + 20),
                cv::FONT_HERSHEY_SIMPLEX,
                0.6,
                cv::Scalar(255, 255, 255),
                1,
                cv::LINE_AA
            );
        }

        // Draw image
        cv::Mat img = images[i];
        if (img.channels() == 1) {
            cv::cvtColor(img, img, cv::COLOR_GRAY2BGR);
        }

        // Resize if needed
        if (img.cols != max_w || img.rows != max_h) {
            float scale = std::min(
                static_cast<float>(max_w) / img.cols,
                static_cast<float>(max_h) / img.rows
            );
            cv::resize(img, img, cv::Size(), scale, scale, cv::INTER_LINEAR);
        }

        // Copy to output
        cv::Rect roi(
            x + (cell_w - img.cols) / 2,
            y + title_height + (max_h - img.rows) / 2,
            img.cols,
            img.rows
        );
        img.copyTo(output(roi));

        // Draw border
        cv::rectangle(
            output,
            cv::Point(x, y),
            cv::Point(x + cell_w - 1, y + cell_h - 1),
            cv::Scalar(80, 80, 80),
            1
        );
    }

    return output;
}

} // namespace utils
} // namespace ivit
